File: packages/server-aspnet/Session.cs
using System.Security.Cryptography;
using System.Text;

namespace OpenPasskey.AspNet;

public class SessionConfig
{
    public string Secret { get; set; } = "";
    public int DurationSeconds { get; set; } = 86400; // 24h
    public int ClockSkewGraceMs { get; set; } = 10_000; // 10s
    public string CookieName { get; set; } = "op_session";
    public string CookiePath { get; set; } = "/";
    public bool Secure { get; set; } = true;
    public string SameSite { get; set; } = "Lax";
    public string? Domain { get; set; }
}

/// <summary>Internal only — never serialized to HTTP responses.</summary>
public record SessionTokenData(string UserId, long ExpiresAt);

public static class SessionHelper
{
    private const int MinSecretLength = 32;

    public static void Validate(SessionConfig config)
    {
        if (string.IsNullOrEmpty(config.Secret) || config.Secret.Length < MinSecretLength)
            throw new ArgumentException($"session secret must be at least {MinSecretLength} characters");
    }

    private static string Sign(string payload, string secret)
    {
        using var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(secret));
        var sig = hmac.ComputeHash(Encoding.UTF8.GetBytes(payload));
        return Convert.ToBase64String(sig).TrimEnd('=').Replace('+', '-').Replace('/', '_');
    }

    public static string CreateToken(string userId, SessionConfig config)
    {
        var expiresAt = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() + (long)config.DurationSeconds * 1000;
        var payload = $"{userId}:{expiresAt}";
        var signature = Sign(payload, config.Secret);
        return $"{payload}:{signature}";
    }

    public static SessionTokenData ValidateToken(string token, SessionConfig config)
    {
        var lastColon = token.LastIndexOf(':');
        if (lastColon == -1) throw new ArgumentException("invalid session token");

        var secondLastColon = token.LastIndexOf(':', lastColon - 1);
        if (secondLastColon == -1) throw new ArgumentException("invalid session token");

        var userId = token[..secondLastColon];
        var expiresAtStr = token[(secondLastColon + 1)..lastColon];
        var providedSig = token[(lastColon + 1)..];

        if (string.IsNullOrEmpty(userId) || string.IsNullOrEmpty(expiresAtStr) || string.IsNullOrEmpty(providedSig))
            throw new ArgumentException("invalid session token");

        if (!long.TryParse(expiresAtStr, out var expiresAt))
            throw new ArgumentException("invalid session token");

        var payload = $"{userId}:{expiresAtStr}";
        var expectedSig = Sign(payload, config.Secret);

        if (!CryptographicOperations.FixedTimeEquals(
                Encoding.UTF8.GetBytes(providedSig),
                Encoding.UTF8.GetBytes(expectedSig)))
            throw new ArgumentException("invalid session token");

        var nowMs = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
        if (nowMs > expiresAt + config.ClockSkewGraceMs)
            throw new ArgumentException("session expired");

        return new SessionTokenData(userId, expiresAt);
    }

    public static string BuildSetCookieHeader(string token, SessionConfig config)
    {
        var parts = new List<string>
        {
            $"{config.CookieName}={token}",
            "HttpOnly",
            $"Path={config.CookiePath}",
            $"Max-Age={config.DurationSeconds}",
            $"SameSite={config.SameSite}"
        };
        if (config.Secure) parts.Add("Secure");
        if (!string.IsNullOrEmpty(config.Domain)) parts.Add($"Domain={config.Domain}");
        return string.Join("; ", parts);
    }

    public static string BuildClearCookieHeader(SessionConfig config)
    {
        var parts = new List<string>
        {
            $"{config.CookieName}=",
            "HttpOnly",
            $"Path={config.CookiePath}",
            "Max-Age=0",
            $"SameSite={config.SameSite}"
        };
        if (config.Secure) parts.Add("Secure");
        if (!string.IsNullOrEmpty(config.Domain)) parts.Add($"Domain={config.Domain}");
        return string.Join("; ", parts);
    }

    public static string? ParseCookieToken(string? cookieHeader, SessionConfig config)
    {
        if (string.IsNullOrEmpty(cookieHeader)) return null;
        var prefix = $"{config.CookieName}=";
        foreach (var cookie in cookieHeader.Split(';'))
        {
            var trimmed = cookie.Trim();
            if (trimmed.StartsWith(prefix))
            {
                var value = trimmed[prefix.Length..];
                return string.IsNullOrEmpty(value) ? null : value;
            }
        }
        return null;
    }
}

File: packages/server-spring/src/main/java/com/openpasskey/spring/Session.java
package com.openpasskey.spring;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.util.Base64;

public class Session {

    private static final int MIN_SECRET_LENGTH = 32;
    private static final long DEFAULT_DURATION_SECONDS = 86400; // 24h
    private static final long DEFAULT_CLOCK_SKEW_GRACE_MS = 10_000; // 10s
    private static final String DEFAULT_COOKIE_NAME = "op_session";

    public record SessionConfig(
            String secret,
            long durationSeconds,
            long clockSkewGraceMs,
            String cookieName,
            String cookiePath,
            boolean secure,
            String sameSite,
            String domain
    ) {
        public SessionConfig {
            if (durationSeconds < 0) durationSeconds = DEFAULT_DURATION_SECONDS;
            if (clockSkewGraceMs < 0) clockSkewGraceMs = DEFAULT_CLOCK_SKEW_GRACE_MS;
            if (cookieName == null || cookieName.isEmpty()) cookieName = DEFAULT_COOKIE_NAME;
            if (cookiePath == null || cookiePath.isEmpty()) cookiePath = "/";
            if (sameSite == null || sameSite.isEmpty()) sameSite = "Lax";
        }

        public SessionConfig(String secret) {
            this(secret, DEFAULT_DURATION_SECONDS, DEFAULT_CLOCK_SKEW_GRACE_MS, DEFAULT_COOKIE_NAME, "/", true, "Lax", null);
        }
    }

    /** Internal only — never serialized to HTTP responses */
    public record SessionTokenData(String userId, long expiresAt) {}

    public static void validate(SessionConfig config) {
        if (config.secret() == null || config.secret().length() < MIN_SECRET_LENGTH) {
            throw new IllegalArgumentException("session secret must be at least " + MIN_SECRET_LENGTH + " characters");
        }
    }

    private static String sign(String payload, String secret) {
        try {
            Mac mac = Mac.getInstance("HmacSHA256");
            mac.init(new SecretKeySpec(secret.getBytes(StandardCharsets.UTF_8), "HmacSHA256"));
            byte[] sig = mac.doFinal(payload.getBytes(StandardCharsets.UTF_8));
            return Base64.getUrlEncoder().withoutPadding().encodeToString(sig);
        } catch (Exception e) {
            throw new RuntimeException("HMAC signing failed", e);
        }
    }

    public static String createToken(String userId, SessionConfig config) {
        long expiresAt = System.currentTimeMillis() + config.durationSeconds() * 1000;
        String payload = userId + ":" + expiresAt;
        String signature = sign(payload, config.secret());
        return payload + ":" + signature;
    }

    public static SessionTokenData validateToken(String token, SessionConfig config) {
        // Split from right — userId may contain colons
        int lastColon = token.lastIndexOf(':');
        if (lastColon == -1) throw new IllegalArgumentException("invalid session token");

        int secondLastColon = token.lastIndexOf(':', lastColon - 1);
        if (secondLastColon == -1) throw new IllegalArgumentException("invalid session token");

        String userId = token.substring(0, secondLastColon);
        String expiresAtStr = token.substring(secondLastColon + 1, lastColon);
        String providedSig = token.substring(lastColon + 1);

        if (userId.isEmpty() || expiresAtStr.isEmpty() || providedSig.isEmpty()) {
            throw new IllegalArgumentException("invalid session token");
        }

        long expiresAt;
        try {
            expiresAt = Long.parseLong(expiresAtStr);
        } catch (NumberFormatException e) {
            throw new IllegalArgumentException("invalid session token");
        }

        // Timing-safe comparison
        String payload = userId + ":" + expiresAtStr;
        String expectedSig = sign(payload, config.secret());

        if (!MessageDigest.isEqual(
                providedSig.getBytes(StandardCharsets.UTF_8),
                expectedSig.getBytes(StandardCharsets.UTF_8))) {
            throw new IllegalArgumentException("invalid session token");
        }

        // Expiry with clock skew grace
        if (System.currentTimeMillis() > expiresAt + config.clockSkewGraceMs()) {
            throw new IllegalArgumentException("session expired");
        }

        return new SessionTokenData(userId, expiresAt);
    }

    public static String buildSetCookieHeader(String token, SessionConfig config) {
        StringBuilder sb = new StringBuilder();
        sb.append(config.cookieName()).append("=").append(token);
        sb.append("; HttpOnly");
        sb.append("; Path=").append(config.cookiePath());
        sb.append("; Max-Age=").append(config.durationSeconds());
        sb.append("; SameSite=").append(config.sameSite());
        if (config.secure()) {
            sb.append("; Secure");
        }
        if (config.domain() != null && !config.domain().isEmpty()) {
            sb.append("; Domain=").append(config.domain());
        }
        return sb.toString();
    }

    public static String buildClearCookieHeader(SessionConfig config) {
        StringBuilder sb = new StringBuilder();
        sb.append(config.cookieName()).append("=");
        sb.append("; HttpOnly");
        sb.append("; Path=").append(config.cookiePath());
        sb.append("; Max-Age=0");
        sb.append("; SameSite=").append(config.sameSite());
        if (config.secure()) {
            sb.append("; Secure");
        }
        if (config.domain() != null && !config.domain().isEmpty()) {
            sb.append("; Domain=").append(config.domain());
        }
        return sb.toString();
    }

    public static String parseCookieToken(String cookieHeader, SessionConfig config) {
        if (cookieHeader == null || cookieHeader.isEmpty()) return null;
        String prefix = config.cookieName() + "=";
        for (String cookie : cookieHeader.split(";")) {
            String trimmed = cookie.trim();
            if (trimmed.startsWith(prefix)) {
                String value = trimmed.substring(prefix.length());
                return value.isEmpty() ? null : value;
            }
        }
        return null;
    }
}
